### PR TITLE
[WIP] Where IN Array support

### PR DESCRIPTION
**Pull request by sdogruyol**
Work in progress. Don't merge yet.

### CODE REVIEW

File: spec/active_record_spec.cr [hunk 1]
```diff
@@ -36,8 +36,23 @@ module ActiveRecord
...
+      low = params["1"] as Int
+      high = params["2"] as Int
+      actual > low && actual < high
+    end
+  end
+
   NullAdapter.register_query("(number_of_dependents > :1) AND (number_of_dependents < :2)",
     LessAndMoreDependents.new)
+
+  NullAdapter.register_query("number_of_dependents IN ((:1,:2))", DependentsIn.new)
```
Comment (sdogruyol) [inline spec/active_record_spec.cr]: The spec errors here. IDK but doesnt register the query.

`Unregistered query for NullAdapter: "(number_of_dependents) IN ((:1, :2))", use NullAdapter.register_query`

Comment (sdogruyol) [inline spec/active_record_spec.cr]: I've found it :smile: The extra whitespace was necessary in query `(number_of_dependents) IN ((:1, :2))`

**Comment by sdogruyol** [comment]
@waterlink 

The specs pass on Crystal `0.14.2` but seems to fail in an unrelated place in `0.15.0`

**Comment by sdogruyol** [comment]
The specs passed but actually it's not working in a real app with Postgresql. It's related to query building with params so i converted the `Array(T)` generator like below.

``` crystal
def _generate(query : ::Array(T), param_count = 0)
  query = query.map do |value|
     "#{value}"
  end.join(", ")
  # This is later being sent to _generate(query : ::ActiveRecord::Criteria, param_count = 0)
  Query.new("(#{query})")
end
```

**Comment by waterlink** [comment]
Injecting values there directly is very "bad idea". Expect SQL injections.

Failures with Postgres adapter should be fixed by overriding the `_generate(...)` for special cases, that need different query generation on Postgres in the `postgres_adapter.cr`.

**Comment by waterlink** [comment]
Oh, now it fails on `0.15.0` because we can't override now behavior of `!`. So we simply have to use `(...).not` now instead of `!(...)`

**Comment by waterlink** [comment]
I will upgrade it to `0.15.0` in separate branch and make a release for it, when everything passes. After that you should rebase your commits on top of updated master branch

**Comment by sdogruyol** [comment]
@waterlink i strongly agree on that :+1: For now i'm still getting the grasp of how this all works.

**Comment by waterlink** [comment]
Yeah, postgres adapter actually already does some overriding there: https://github.com/waterlink/postgres_adapter.cr/blob/master/src/postgres_adapter.cr#L10-L15

The same overload for `Array(T)` is required I assume.

**Comment by sdogruyol** [comment]
@waterlink actually the problem is that `in([1,2,3])` is turned into `Criteria` and then later the params is not being merged into the query. 

Example.

`SELECT * FROM item_settings WHERE item_id IN (:1, :2, :3, :4)`

This gives SQL error since the syntax is not correct.

**Comment by waterlink** [comment]
And maybe it makes sense to parametrize the code by the parameter name transformator and extract it to class of its own and re-use it from all implementations. Something like that:

``` crystal
# in active_record
class ArrayQueryHandler
  def initialize(@param_name_transformer : Int32 -> String)
  end

  def handle(query : Array(T), param_count = 0)
    params = {} of String => T
    query = query.map do |value|
      param_count += 1
      params[param_count.to_s] = value
      @param_name_transformer.call(param_count)
    end.join(", ")
    {Query.new("(#{query})", params), param_count}
  end
end

# and usage in active_record:
def _generate(query : Array(T), param_count = 0)
  result, param_count = ArrayQueryHandler.new { |name| ":#{name}" }.handle(query)
  result
end

# and usage in prosgres_adapter
def _generate(query : Array(T), param_count = 0)
  result, param_count = ArrayQueryHandler.new { |name| "$#{name}" }.handle(query)
  result
end

# see closely at the difference in the block.
```

**Comment by waterlink** [comment]
That sounds weird, it should not be converted to criteria. Could you update the branch (rebase against upstream master) and push it to github? So I can fetch it and play around to see what is happening.

**Comment by sdogruyol** [comment]
@waterlink updated

**Comment by sdogruyol** [comment]
I've add `ArrayQueryHandler` like you said :+1:

**Comment by sdogruyol** [comment]
So just like you said adding below to `postgres_adapter` does it :tada: 

``` crystal
def _generate(query : Array(T), param_count = 0)
    result, param_count = ::ActiveRecord::Sql::ArrayQueryHandler.new { |name| "$#{name}" }.handle(query)
    result
end
```

Opening a PR now to `postgres_adapter` :smile:

**Comment by waterlink** [comment]
Great! I am going to merge this one then.

**Comment by sdogruyol** [comment]
Thanks a lot :heart_eyes:

**Pull request merged**